tests/rhbz_functions.at

```diff
@@ -0,0 +1,275 @@
+            xmlrpc_value *component_struct = NULL;
+            xmlrpc_value *sub_components_list = NULL;
+
+            xmlrpc_decompose_value(&env, result, "{s:(S),*}", "products", &product_struct);
+            if (env.fault_occurred)
+                printf("Error retrieving product_struct: %s\n", env.fault_string);
+
+            xmlrpc_struct_find_value(&env, product_struct, "components", &components_list);
+            if (env.fault_occurred)
+                printf("Error retrieving components_list: %s\n", env.fault_string);
+
+            unsigned size = xmlrpc_array_size(&env, components_list);
+            for (unsigned i = 0; i < size; i++)
+            {
+                const char *name = NULL;
+                xmlrpc_array_read_item(&env, components_list, i, &component_struct);
+                if (env.fault_occurred)
+                    printf("Error reading component_struct from components_list: %s\n", env.fault_string);
+                name = read_string(&env, component_struct, "name");
+                if (strcmp(name, "gdb") == 0)
+                {
+                    free((void *)name);
+                    break;
+                }
+                free((void *)name);
+                xmlrpc_DECREF(component_struct);
+            }
+            xmlrpc_struct_find_value(&env, component_struct, "sub_components", &sub_components_list);
+            if (env.fault_occurred)
+                printf("Error retrieving sub_components_list: %s\n", env.fault_string);
+
+            TS_ASSERT_PTR_IS_NOT_NULL(sub_components_list);
+
+            xmlrpc_DECREF(result);
+            xmlrpc_DECREF(product_struct);
+            xmlrpc_DECREF(component_struct);
+            xmlrpc_DECREF(components_list);
+            if (sub_components_list)
+                xmlrpc_DECREF(sub_components_list);
+        }
+
+        {
+            problem_data_t *problem_data = create_problem_data_for_reporting("../../sample_problem");
+            bug_id = rhbz_new_bug(client,
+                problem_data, "Red Hat Enterprise Linux 8", "8.0",
+                "Foo doesn't work",
+                "It's been broken for ages",
+                rhbz.b_create_private,
+                rhbz.b_private_groups
+                );
+        }
+
+        {
+            xmlrpc_value *sub_components = rhbz_get_sub_components(client, "Red Hat Enterprise Linux 8", "afmw");
+            TS_ASSERT_PTR_IS_NULL(sub_components);
+            if (sub_components)
+                xmlrpc_DECREF(sub_components);
+        }
+
+        {
+            xmlrpc_value *sub_components = rhbz_get_sub_components(client, "Fedora", "afmw");
+            TS_ASSERT_PTR_IS_NULL(sub_components);
+            if (sub_components)
+                xmlrpc_DECREF(sub_components);
+        }
+
+        /* Should succeed even though we don't want to report sub_components for Fedora
+         * because the product name check is up in rhbz_new_bug. Do we want to move it one level down
```

Comment:
I don’t think we do. The result would be, what, an empty value? That sounds fine to me.